### PR TITLE
reduce step down to i8

### PR DESCRIPTION
**Pull request by z8674558**
Closes https://github.com/internet2-org/rust-aluvm/issues/43

### CODE REVIEW

**Comment by z8674558** [comment]
This approach has two drawbacks:

- Bytecode looks a bit ugly
    - | `reg` (3 bit) | `idx` (5 bit) | `add/sub flag` (it is actually 1 bit) | `step size` (8 bit) |
    - if we split `stp` operation to `add` and `sub`, we can store the code in 3 bytes
- This makes hard to apply a big value to registers like `A128` with `stp`
    - workaround is to use two registers and make use of `mul`

**Comment by z8674558** [comment]
This issue is not significant though. We can leave it for now if we do not imagine use-cases

**Comment by dr-orlovsky** [comment]
Hm, we can't introduce new instructions since it will break existing instruction set (all values are already allocated). We also can't have non-used bits in the bytecode per AluVM policy. So the only option here is to use `i8` instead of `u8` for a step value - and do not use flag. This will fit into 3 bytes. 

It must not affect `PutA` anyhow; the only thing which will be affected are `add/sub int,reg` instructions - but it is fine since we always can do `put bigint, reg1; add reg1, reg2` for large integers (this is ok for RISC command sets).

**Comment by z8674558** [comment]
makes sense. will work on

**Comment by z8674558** [comment]
I noticed that currently `reshape` work like following:

`[255]` (-1 represented by little endian 1 byte)

↓ (reshape to 2 byte)

`[255, 0]` (this is not -1)

I think we should return `[255, 255]` in this case. Fixed in https://github.com/internet2-org/rust-aluvm/pull/45

**Comment by dr-orlovsky** [comment]
@z8674558 BTW are you aware of https://github.com/pandoracore/aluasm project which makes it easier to use and debug AluVM (you can write assembly as files, not rust macros, and compile them + use libraries).

Also AluRE (Alu runtime environment) moves this further and provides runtime environment - but it is very early stage right now: https://github.com/pandoracore/AluRE

Finally there is a concept of a high-level language for AluVM which may be used for smart contracts: https://github.com/rgb-org/contractum-lang

**Comment by z8674558** [comment]
cool, `Contructum` catches my eye. It looks `AluRE` is a private repo? :)

**Comment by dr-orlovsky** [comment]
Oh, it was (since there actually nothing to see yet). Opened it.

File: src/isa/exec.rs [hunk 1]
```diff
@@ -470,7 +470,9 @@ impl InstructionSet for ArithmeticOp {
                 idx,
                 regs.get(reg, idx).and_then(|val| {
                     let mut n = Number::from(*step);
-                    n.reshape(val.layout());
+                    if !n.reshape(val.layout()) {
+                        unreachable!("reshape target byte length is always greater")
+                    }
```
Comment (dr-orlovsky) [inline src/isa/exec.rs:475]: good check!

BTW you can use more simple form
`debug_assert!(n.reshape(val.layout()), "reshape target byte length is always greater");`

Using `debug_assert` allows to get more efficient code in releases (they are cut by compiler for release profiles, avoiding unnecessary jumps in the code)

File: src/libs/cursor.rs [hunk 2]
```diff
@@ -256,6 +256,17 @@ where
...
     }
 
+    fn read_i8(&mut self) -> Result<i8, CodeEofError> {
+        if self.is_eof() {
+            return Err(CodeEofError);
+        }
+        let pos = self.byte_pos as usize;
+        let mut buf = [0u8];
+        buf.copy_from_slice(&self.as_ref()[pos..pos + 1]);
+        let byte = i8::from_le_bytes(buf);
```
Comment (dr-orlovsky) [inline src/libs/cursor.rs:266]: Can be simplified as

```suggestion
        let byte = self.as_ref()[self.byte_pos as usize] as i8;
```

File: src/libs/cursor.rs [hunk 3]
```diff
@@ -382,6 +393,13 @@ where
         self.inc_bytes(1).map_err(WriteError::from)
     }
 
+    fn write_i8(&mut self, data: impl Into<i8>) -> Result<(), WriteError> {
+        let data = data.into().to_le_bytes();
+        let pos = self.byte_pos as usize;
+        self.as_mut()[pos] = data[0];
```
Comment (dr-orlovsky) [inline src/libs/cursor.rs:399]: Just

```suggestion
        let pos = self.byte_pos as usize;
        self.as_mut()[pos] = data.into();
```

**Review by dr-orlovsky — COMMENTED** [review]
Minor improvements, otherwise LGTM

File: src/isa/exec.rs (hunk 1 above)
Comment (z8674558) [reply src/isa/exec.rs]: I was not aware of it. thanks

File: src/libs/cursor.rs (hunk 3 above)
Comment (z8674558) [reply src/libs/cursor.rs:396]: With my understanding we cannot do that because `data` is `impl Into<i8>` while self is `AsMut<[u8]>`

Comment (dr-orlovsky) [reply src/libs/cursor.rs:396]: Ok, agree with your version then

**Review by dr-orlovsky — APPROVED** [review]
ACK 832bbc06ef3d31533711855b7d66061456a0e0ff

**Pull request merged**